### ISSUE TITLE
Listening to stream as EventSource

### ISSUE THREAD

**Issue by hajsf** [grpc-ecosystem/grpc-gateway#2786]
This is not a bug, but a clarification.

I built my gateway that is sending a stream very well to the 'postman/Thunder client' as `http://localhost:8090/v1/example/stream?id=5`

<img width="1076" alt="image" src="https://user-images.githubusercontent.com/98168280/178351947-cc53418d-c0ee-4286-9aeb-a299c6496160.png">

I would like to recieve the stream at my web page, so thought the best way is to use `EvenSource` as:
```JavaScript
    const eventSource = new EventSource("http://localhost:8090/v1/example/stream?id=5");
    eventSource.onmessage = function (event) {
        console.log(event)
    }
```
But I'm getting error:
```bash
EventSource's response has a MIME type ("application/json") that is not "text/event-stream". Aborting the connection.
```

And if I tried calling it via `fetch` function, as:
```JavaScript
    window.onload = function() {
        var apiUrl = 'http://localhost:8090/v1/example/stream?id=5';
        fetch(apiUrl)
            .then(response => {
                console.log(response)
                var data= JSON.parse(response);
                return data.json()})
            .then(data => console.log(data))
            .catch(err => { console.log(err)});
    }; 
```
Then I'm getting:
1. One responce only,
2. Can not parse the response to JSON, and get error at ``var data= JSON.parse(response);`

<img width="815" alt="image" src="https://user-images.githubusercontent.com/98168280/178354892-d6b34d0c-9180-462d-b28a-20d59d09864a.png">

I tried changing the header to be:
```go
	md := metadata.New(map[string]string{"Content-Type": "text/event-stream", "Connection": "keep-alive"})
	srv.SetHeader(md)
```
But no use. below is my full code:
```go
func (s server) FetchResponse(in *pb.Request, srv pb.StreamService_FetchResponseServer) error {
	if p, ok := peer.FromContext(srv.Context()); ok {
		fmt.Println("Client ip is:", p.Addr.String())
	}
	md := metadata.New(map[string]string{"Content-Type": "text/event-stream", "Connection": "keep-alive"})
	srv.SetHeader(md)
	//use wait group to allow process to be concurrent
	var wg sync.WaitGroup
	for i := 0; i < 5; i++ {
		wg.Add(1)
		go func(count int64) {
			defer wg.Done()
			//w.Header().Set("Content-Type", "text/event-stream")
			//time sleep to simulate server process time
			time.Sleep(time.Duration(count) * time.Second)
			resp := pb.Response{Result: fmt.Sprintf("Request #%d For Id:%d", count, in.Id)}
			if err := srv.Send(&resp); err != nil {
				log.Printf("send error %v", err)
			}
			log.Printf("finishing request number : %d", count)
		}(int64(i))
	}

	wg.Wait()
	return nil
}
```

**Comment by johanbrandhorst** [comment]
Hi, thanks for your issue! I don't know if anyone has successfully used the grpc-gateway for event sourcing before, this is exciting 😄. For configuring the content type response, I think your best bet is to create a custom marshaler and register it:

```go
type eventSourceMarshaler struct {
	runtime.JSONPb
}

func (m *eventSourceMarshaler) ContentType() string {
	return "text/event-stream"
}


func main() {
	mux := runtime.NewServeMux(runtime.WithMarshalerOption("text/event-stream", &eventSourceMarshaler{JSONPb: runtime.JSONPb{}}))
}
```

Something like this might work?

**Comment by hajsf** [comment]
> ```go
> &eventSourceMarshaler{JSONPb: runtime.JSONPb{}}
> ```

I got:
```bash
cannot use &(eventSourceMarshaler literal) (value of type *eventSourceMarshaler) as "github.com/grpc-ecosystem/grpc-gateway/v2/runtime".Marshaler value in argument to runtime.WithMarshalerOption: *eventSourceMarshaler does not implement "github.com/grpc-ecosystem/grpc-gateway/v2/runtime".Marshaler (wrong type for method ContentType)
		have ContentType() string
		want ContentType(v interface{}) 
```

**Comment by johanbrandhorst** [comment]
No results? So you mean you're still seeing the same error in your frontend? Are you sending the `Accept: text/event-stream` header in the request? Could could continue debugging by adding print statements to the marshaler to see that it's being called.

**Comment by MikeSouza** [comment]
@hajsf Curious if adding the custom marshaller ended up working for you?

**Comment by mgoltzsche** [comment]
I ran into a similar problem: I noticed that the server is buffering streamed responses when the client sends the HTTP header `Accept-Encoding: gzip, deflate, br, zstd` which however the browser forcefully (cannot be overwritten by the JS app) injects into requests when using the `fetch` JS API, breaking the HTTP stream in some cases.
Now I wonder whether the browser wouldn't send that header when using the `EventSource` JS API instead of `fetch`.
@hajsf can you please confirm whether adding the `Accept: text/event-stream` header made it work for you and whether the response is then really streamed and not sent at all at once only after the stream finished?

**Comment by mgoltzsche** [comment]
I created a [simple example](https://gist.github.com/mgoltzsche/17ddd2b878a09524c32191c2768036f2) to test the `EventSource` JS API and answer some of the questions we have here:
* Using the `EventSource` JS API doesn't prevent the browser from sending the `Accept-Encoding: gzip, deflate, br, zstd` header - however Golang's `net/http` library is still not buffering the response.
* The `EventSource` JS API does not only expect the Content-Type `text/event-stream` but it also expects the response body to be formatted correspondingly: it is not a JSON stream but a text stream where events are separated by two new lines (`\n\n`) and each event line must start with the type of event separated by a colon and an actual message event must start with `data: `, followed by the actual message content which may as well be a JSON object. Though, that format could be implemented using a custom marshaller but the ones listed in this issue are not cutting it. (Personally I don't care whether `fetch` or `EventSource` has to be used but I just want to get rid of the buffering which unfortunately affects both approaches :disappointed: .)

**Comment by johanbrandhorst** [comment]
Sorry if I'm missing something, were you able to pinpoint where the buffering is happening? Is there anything we can do in the gateway?

**Comment by mgoltzsche** [comment]
yes, indeed, grpc-gateway didn't cause the buffering but somebody recently added a gzip compressing middleware in front of the stream in our code base :sweat_smile:.

**Comment by mgoltzsche** [comment]
fyi: here's the working `text/event-stream` marshaller (with the correct body format) I used for experimentation at some point:
```go
type eventSourceMarshaler struct {
	grpc_runtime.JSONPb
}

func (m *eventSourceMarshaler) ContentType(v interface{}) string {
	return "text/event-stream"
}

func (m *eventSourceMarshaler) Marshal(v interface{}) ([]byte, error) {
	b, err := m.JSONPb.Marshal(v)
	if err != nil {
		return nil, err
	}

	return []byte(fmt.Sprintf("data: %s\n", string(b))), nil
}

func (m *eventSourceMarshaler) Unmarshal(data []byte, v interface{}) error {
	return errors.New("text/event-stream unmarshaler is unsupported")
}
```

**Comment by johanbrandhorst** [comment]
Awesome, would you be interested in writing something for our docs for others who want to use Event sourcing like this? Then we can close this issue.

**Comment by mgoltzsche** [comment]
Not that much :smile:. Also, I just quickly hacked that together and tested it for experimentation purposes. While doing so I was wondering why the Marshaler inteface requires both the `Marshal` and `Unmarshal` methods to be implemented in addition to the `NewEncoder` and `NewDecoder` methods. Apparently the last two are not used which is why I didn't overwrite them in my implementation but let it inherit the default implementation of those methods. Are those methods actually used somewhere or is the inteface just a bit cluttered?

**Comment by johanbrandhorst** [comment]
I'm actually not sure, but I've only ever implemented Marshal and Unmarshal 😅
